### PR TITLE
1304 refactor certificate domain

### PR DESCRIPTION
**Pull request by ckr123**

### CODE REVIEW

File: domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs [hunk 1]
```diff
@@ -14,17 +15,25 @@
 
 namespace API.AppTests;
 
-public sealed class CertificateIssuingTests : IClassFixture<QueryApiWebApplicationFactory>, IClassFixture<MartenDbContainer>, IDisposable
+public sealed class CertificateIssuingTests :
+    IClassFixture<QueryApiWebApplicationFactory>,
+    IClassFixture<MartenDbContainer>,
+     IClassFixture<RabbitMqContainer>,
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs:21]: Minor thing: Indentation is a bit funny here

File: domains/certificates/Query.API/API/RabbitMq/Configurations/RabbitMqOptions.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+namespace API.RabbitMq.Configurations;
+
+public class RabbitMqOptions
+{
+    public const string RabbitMq = "RabbitMq";
+
+    public string Password { get; set; } = string.Empty;
+    public string Username { get; set; } = string.Empty;
+    public string Host { get; set; } = string.Empty;
+    public string Port { get; set; } = string.Empty;
```
Comment (duizer) [inline domains/certificates/Query.API/API/RabbitMq/Configurations/RabbitMqOptions.cs:10]: Port should be changed to a integer (or maybe a ushort) in stead of a string.

File: domains/certificates/Query.API/API/appsettings.Development.json [hunk 3]
```diff
@@ -10,5 +10,11 @@
     },
     "Datasync": {
         "Url": "http://data-sync-api"
+    },
+    "RabbitMq": {
+        "Host": "localhost",
+        "Port": "5672",
+        "Username": "guest",
+        "Password": "guest"
```
Comment (duizer) [inline domains/certificates/Query.API/API/appsettings.Development.json:18]: What about our docker-compose setup? It is not part of this PR, so I guess that these values will only work if you manually start a rabbitmq instance.

File: domains/certificates/Query.API/API.AppTests/SwaggerTests.cs [hunk 4]
```diff
@@ -9,11 +10,18 @@
 namespace API.AppTests;
 
 [UsesVerify]
-public class SwaggerTests : IClassFixture<QueryApiWebApplicationFactory>
+public class SwaggerTests : IClassFixture<QueryApiWebApplicationFactory>, IClassFixture<RabbitMqContainer>
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/SwaggerTests.cs:13]: Is the RabbitMqContainer needed in order for the swagger tests to work? If yes, is that due to the WaitUntilStarted trick we did in QueryApiWebApplicationFactory ?

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            {
+                Username = "guest",
+                Password = "guest"
+            })
+            .Build();
+
+    public string Username => testContainer.Username;
+    public string Password => testContainer.Password;
+
+
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:25]: Minor thing: These spaces can be deleted

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+public class RabbitMqContainer : IAsyncLifetime
+{
+    private readonly RabbitMqTestcontainer testContainer;
+
+    public RabbitMqContainer() =>
+        testContainer = new TestcontainersBuilder<RabbitMqTestcontainer>()
+            .WithMessageBroker(new RabbitMqTestcontainerConfiguration
+            {
+                Username = "guest",
+                Password = "guest"
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:18]: Since username and password is exposed as properties, is it necessary to set them here? Will they be null otherwise?

File: domains/certificates/Query.API/API.AppTests/ContractTests.cs [hunk 7]
```diff
@@ -5,22 +5,27 @@
...
+    public ContractTests(QueryApiWebApplicationFactory factory, MartenDbContainer marten, RabbitMqContainer rabbitMqContainer)
     {
         dataSyncWireMock = new DataSyncWireMock(port: 9001);
         this.factory = factory;
         this.factory.MartenConnectionString = marten.ConnectionString;
         this.factory.DataSyncUrl = dataSyncWireMock.Url;
+        this.factory.RabbitMqHost = rabbitMqContainer.Hostname;
+        this.factory.RabbitMqPort = rabbitMqContainer.Port.ToString();
+        this.factory.RabbitMqUsername = rabbitMqContainer.Username;
+        this.factory.RabbitMqPassword = rabbitMqContainer.Password;
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/ContractTests.cs:28]: Suggestion: Instead of having 4 properties to set on factory we can add a method that takes all 4 arguments. This ensures that someone does not forget to set 1 of the 4 properties.

File: domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs [hunk 8]
```diff
@@ -14,17 +15,25 @@
...
+    IDisposable
 {
     private readonly QueryApiWebApplicationFactory factory;
     private readonly DataSyncWireMock dataSyncWireMock;
 
-    public CertificateIssuingTests(QueryApiWebApplicationFactory factory, MartenDbContainer martenDbContainer)
+    public CertificateIssuingTests(QueryApiWebApplicationFactory factory, MartenDbContainer martenDbContainer, RabbitMqContainer rabbitMqContainer)
     {
-        dataSyncWireMock = new DataSyncWireMock(port: 9002);
+        dataSyncWireMock = new DataSyncWireMock(port: 9003);
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs:29]: Side note: I remember now why I set this to 9002. ContractTests uses port 9001. Since it is possible that CertificateIssuingTests and ContractTests execute in parallel, then the port should unique for the test class.
I will try to see if there is a way to both randomize the port and check if the port is open in Wiremock

File: domains/certificates/Query.API/API/Program.cs [hunk 9]
```diff
@@ -57,7 +57,7 @@
 
 builder.Services.AddHealthChecks();
 
-builder.Services.AddIntegrationEventBus();
+builder.Services.AddRabbitMq(builder.Configuration);
```
Comment (duizer) [inline domains/certificates/Query.API/API/Program.cs:60]: Will the updated diagrams and documentation be part of this PR?

**Review by duizer — CHANGES_REQUESTED** [review]

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+using Xunit;
+
+namespace API.IntegrationTest.Infrastructure;
+
+public class RabbitMqContainer : IAsyncLifetime
+{
+    private readonly RabbitMqTestcontainer testContainer;
+
+    public RabbitMqContainer() =>
+        testContainer = new TestcontainersBuilder<RabbitMqTestcontainer>()
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:14]: Which docker image and version of RabbitMQ will this spin up? Is it possible to set it?
In production, the RabbitMQ version is determined by the RabbitMQ operator. For the version we are using, this will use 
this docker image+tag: "bitnami/rabbitmq:3.10.14-debian-11-r1"

File: domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs [hunk 11]
```diff
@@ -23,14 +23,24 @@ public class QueryApiWebApplicationFactory : WebApplicationFactory<Program>
...
+        builder.UseSetting("RabbitMq:Password", RabbitMqPassword);
+        builder.UseSetting("RabbitMq:Username", RabbitMqUsername);
+        builder.UseSetting("RabbitMq:Host", RabbitMqHost);
+        builder.UseSetting("RabbitMq:Port", RabbitMqPort);
 
         builder.ConfigureTestServices(services =>
         {
+            //  Ensure masstransit bus is started when we run our health checks
+            services.AddOptions<MassTransitHostOptions>().Configure(options => options.WaitUntilStarted = true);
             //Remove DataSyncSyncerWorker
```
Comment (anna-knudsen) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs:44]: Unnecessary comment? (Remove DataSyncSyncerWorker)

File: doc/architecture/domains/certificates.md [hunk 12]
```diff
@@ -17,9 +17,7 @@ The reason for this split is certain constraints. The constraints are:
 
 ## Component diagram: Certificate API
 
-The component diagram shows a first iteration which is based on an in-memory integration event bus. A consequence of using the in-memory implementation is that all components that is dependent on the integration event bus must be in same container.
-
-Components that is used for mocking and will be replaced or discarded at a later are marked with its own color in the diagram.
+The component diagram shows how the solution works, based on a RabbitMQ message broker that publishes the events received from the DataSyncSyncer.
```
Comment (duizer) [inline doc/architecture/domains/certificates.md:20]: In the code changes in the PR, the name "RabbitMQ" is used (domains/certificates/Query.API/API/RabbitMq/Startup.cs), but in the diagram it is called "Message Broker". Before it was called "Integration Event Bus" both in the code and in the diagram. Is it possible that we can have that the name of a component (or container) is named the same in both code and in the diagram?

**Review by duizer — CHANGES_REQUESTED** [review]

File: domains/certificates/Query.API/API/Program.cs [hunk 13]
```diff
@@ -57,7 +57,7 @@
 
 builder.Services.AddHealthChecks();
 
-builder.Services.AddIntegrationEventBus();
```
Comment (duizer) [inline domains/certificates/Query.API/API/Program.cs:60]: We have a shared library named "IntegrationEvents" with events we expected to receive on the Integration Event Bus. Given that we here delete the concept "Integration Event Bus" from the code, should we not also do something with event(s) under the library "IntegrationEvents"?

File: domains/certificates/Query.API/API.AppTests/HealthTests.cs [hunk 14]
```diff
@@ -1,17 +1,22 @@
 using System.Net;
 using System.Threading.Tasks;
 using API.AppTests.Infrastructure;
+using API.IntegrationTest.Infrastructure;
 using Xunit;
 
 namespace API.AppTests;
 
-public class HealthTests : IClassFixture<QueryApiWebApplicationFactory>
+public class HealthTests : IClassFixture<QueryApiWebApplicationFactory>, IClassFixture<RabbitMqContainer>
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/HealthTests.cs:9]: Now that we a way of getting the application in an unhealthy state (by not having a RabbitMQ container running), should we have a test that checks that the application becomes unhealthy?

File: domains/certificates/Query.API/API/appsettings.Development.json [hunk 15]
```diff
@@ -10,5 +10,11 @@
     },
     "Datasync": {
         "Url": "http://data-sync-api"
+    },
+    "RabbitMq": {
+        "Host": "localhost",
+        "Port": "5672",
```
Comment (duizer) [inline domains/certificates/Query.API/API/appsettings.Development.json:16]:
```suggestion
        "Port": 5672,
```

File: domains/certificates/Query.API/API.AppTests/ContractTests.cs (hunk 7 above)
Comment (duizer) [reply domains/certificates/Query.API/API.AppTests/ContractTests.cs:28]: If it is a method on the factory, then that method would have to store the arguments in backing fields, which is then used in the factory's ConfigureTestServices.

A maybe better approach is to expose a single property on the factory of type RabbitMqOptions. This type has all the properties anyway.

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs (hunk 10 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:14]: It will spin up "rabbitmq:3.7.28" - are you suggesting being specific on the image? 
Then we will need to maintain it

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs (hunk 6 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:18]: Yes it is necessary, else testcontainer will look for some env vars that doesnt exists

File: domains/certificates/Query.API/API.AppTests/SwaggerTests.cs (hunk 4 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/SwaggerTests.cs:14]: Yes else it will keep waiting

File: domains/certificates/Query.API/API.AppTests/HealthTests.cs (hunk 14 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/HealthTests.cs:10]: Good point, i do have some troubles getting the health check to act as i want it to act. :) Maybe we can have a chat about it?

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs (hunk 10 above)
Comment (duizer) [reply domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:14]: Yes, I am suggesting that it set to the same image as will be used in production. The same also goes for the docker-compose image for RabbitMQ.
Unfortunately, it is determined by the version of RabbitMQ operator and is defined in eo-base-environment. So it will be a "magic string", but cannot see a way out of solving that right now.

File: domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs (hunk 8 above)
Comment (duizer) [reply domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs:31]: No. But removing the port number inside DataSyncWireMock should mean that it gets a random port assigned. I have not tried if there should be any issues. You are welcome to do so. If you choose not, then close the conversation.

File: domains/certificates/Query.API/API.AppTests/HealthTests.cs (hunk 14 above)
Comment (duizer) [reply domains/certificates/Query.API/API.AppTests/HealthTests.cs:10]: Sure

File: domains/certificates/Shared/RabbitMqEvents/EnergyMeasuredIntegrationEvent.cs [hunk 16]
```diff
@@ -1,4 +1,4 @@
-namespace IntegrationEvents;
+namespace RabbitMqEvents;
```
Comment (duizer) [inline domains/certificates/Shared/RabbitMqEvents/EnergyMeasuredIntegrationEvent.cs:1]: Have you considered other names than RabbitMqEvents? I was thinking something like "MeasurementEvents" or "MeasurementMessages" as this is releated to measurements.

File: domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs (hunk 8 above)
Comment (duizer) [reply domains/certificates/Query.API/API.AppTests/CertificateIssuingTests.cs:31]: I tried just removing the port, but our tests then fails. We/I have to do a bit more investigation to why that is. But that is outside of the scope of this task. I am closing the conversation.

File: domains/certificates/Shared/RabbitMqEvents/EnergyMeasuredIntegrationEvent.cs (hunk 16 above)
Comment (ckr123) [reply domains/certificates/Shared/RabbitMqEvents/EnergyMeasuredIntegrationEvent.cs]: renamed to MeasurementEvents

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs (hunk 10 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:15]: This will introduce some errors with M1 processors

File: domains/certificates/Query.API/API.AppTests/HealthTests.cs (hunk 14 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/HealthTests.cs:10]: As agreed an unhealthtests class has been created

File: domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs [hunk 17]
```diff
@@ -24,24 +25,44 @@ public class QueryApiWebApplicationFactory : WebApplicationFactory<Program>
...
+            {
+                //  Ensure masstransit bus is started when we run our health checks
+                services.AddOptions<MassTransitHostOptions>().Configure(options => options.WaitUntilStarted = true);
+            }
             //Remove DataSyncSyncerWorker
             services.Remove(services.First(s => s.ImplementationType == typeof(DataSyncSyncerWorker)));
         });
     }
 
+    private bool IsRabbitMqSet() => RabbitMqSetup != null;
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs:57]: Could we make the property nullable instead? That would also fix the code smell warning from SonarCloud.

File: domains/certificates/Query.API/API.AppTests/UnhealthTests.cs [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace API.AppTests;
+
+public class UnhealthTests : IClassFixture<QueryApiWebApplicationFactory>
+{
+    private readonly QueryApiWebApplicationFactory factory;
+
+    public UnhealthTests(QueryApiWebApplicationFactory factory)
+    {
+        this.factory = factory;
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/UnhealthTests.cs:14]: If the `RabbitMqSetup` property is nullable, we could here have a line like `this.factory.RabbitMqSetup = null`, which indicates to the reader that there is no RabbitMQ. What do you think?

File: domains/certificates/Query.API/API.AppTests/SwaggerTests.cs [hunk 19]
```diff
@@ -9,11 +11,21 @@
...
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/SwaggerTests.cs:14]: The code has now been changed in the factory to not wait for the bus, if RabbitMqSetup is null. So can we remove the RabbitMqContainer here from this test now?

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    public string Username => testContainer.Username;
+    public string Password => testContainer.Password;
+
+
+    public string Hostname => testContainer.Hostname;
+    public int Port => testContainer.Port;
+
+    public async Task InitializeAsync() => await testContainer.StartAsync();
+
+    public Task DisposeAsync() => testContainer.StopAsync(); //  testContainer.DisposeAsync().AsTask();
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:32]:
```suggestion
    public Task DisposeAsync() => testContainer.StopAsync();
```

File: domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+                Password = "guest"
+            })
+            .Build();
+
+    public string Username => testContainer.Username;
+    public string Password => testContainer.Password;
+
+
+    public string Hostname => testContainer.Hostname;
+    public int Port => testContainer.Port;
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/RabbitMqContainer.cs:28]:
```suggestion
    public string Password => testContainer.Password;
    public string Hostname => testContainer.Hostname;
    public int Port => testContainer.Port;
```

File: domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs [hunk 22]
```diff
@@ -24,24 +25,44 @@ public class QueryApiWebApplicationFactory : WebApplicationFactory<Program>
...
 
+    public RabbitMqOptions RabbitMqSetup { get; set; }
+
+
     protected override void ConfigureWebHost(IWebHostBuilder builder)
     {
         builder.UseSetting("ConnectionStrings:Marten", MartenConnectionString);
         builder.UseSetting("Datasync:Url", DataSyncUrl);
 
+        if (IsRabbitMqSet())
```
Comment (duizer) [inline domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs:36]: I think that as a default the ASPNETCORE_ENVIRONMENT will be set to development. If that is right, the application **may** (I am not sure) use the values listed in appsettings.development.json in QueryAPI-project. 
If not overwritten (e.g. when IsRabbitSet() is false), then will it have null/empty values or use the ones from appsettings.development.json? I am not sure. But if it uses appsettings.development.json values, then someone running the test locally will have strange test results if there is a rabbitmq running.

Doest it make sense?

File: domains/certificates/Query.API/API.AppTests/SwaggerTests.cs (hunk 19 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/SwaggerTests.cs:14]: Seems some of the test still want to spin up rabbitmq. How is the lifecycle if these things?

File: domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs (hunk 22 above)
Comment (ckr123) [reply domains/certificates/Query.API/API.AppTests/Infrastructure/QueryApiWebApplicationFactory.cs:36]: Not sure if it make sense, but no problems here when running a rabbitmq while running the tests

**Review by duizer — APPROVED** [review]
